src/test/java/com/dataartisans/flink/dataflow/streaming/StateSerializationTest.java

```diff
@@ -0,0 +1,242 @@
+		}
+		return comp;
+	}
+
+	private void storeState(StateBackend.CheckpointStateOutputView out) throws Exception {
+		StateCheckpointWriter checkpointBuilder = StateCheckpointWriter.create(out);
+		Coder<String> keyCoder = StringUtf8Coder.of();
+
+		// checkpoint the timers
+		StateCheckpointUtils.encodeTimers(activeTimers, checkpointBuilder,keyCoder);//encodeTimers(out, keyCoder);
+
+		// checkpoint the state
+		StateCheckpointUtils.encodeState(statePerKey, checkpointBuilder, keyCoder);//encodeState(out, keyCoder);
+	}
+
+	private boolean checkTimersForKey(Set<TimerInternals.TimerData> originalTimers, Set<TimerInternals.TimerData> restoredTimers) {
+		boolean comp = true;
+		if(restoredTimers == null) {
+			return false;
+		}
+
+		if(originalTimers.size() != restoredTimers.size()) {
+			return false;
+		}
+
+		for(TimerInternals.TimerData timer: originalTimers) {
+			comp &= restoredTimers.contains(timer);
+		}
+		return comp;
+	}
+
+	private boolean checkStateForKey(FlinkStateInternals<String> originalState, FlinkStateInternals<String> restoredState) throws CannotProvideCoderException {
+		if(restoredState == null) {
+			return false;
+		}
+
+		ValueState<String> orValue = originalState.state(NAMESPACE_1, STRING_VALUE_ADDR);
+		ValueState<String> resValue = restoredState.state(NAMESPACE_1, STRING_VALUE_ADDR);
+		boolean comp = orValue.get().read().equals(resValue.get().read());
+
+		ValueState<Integer> orIntValue = originalState.state(NAMESPACE_1, INT_VALUE_ADDR);
+		ValueState<Integer> resIntValue = restoredState.state(NAMESPACE_1, INT_VALUE_ADDR);
+		comp &= orIntValue.get().read().equals(resIntValue.get().read());
+
+		CombiningValueState<Integer, Integer> combOrValue = originalState.state(NAMESPACE_1, SUM_INTEGER_ADDR);
+		CombiningValueState<Integer, Integer> combResValue = restoredState.state(NAMESPACE_1, SUM_INTEGER_ADDR);
+		comp &= combOrValue.get().read().equals(combResValue.get().read());
+
+		WatermarkStateInternal orWatermark = originalState.state(NAMESPACE_1, WATERMARK_BAG_ADDR);
+		WatermarkStateInternal resWatermark = restoredState.state(NAMESPACE_1, WATERMARK_BAG_ADDR);
+		comp &= orWatermark.get().read().equals(resWatermark.get().read());
+
+		BagState<String> orBag = originalState.state(NAMESPACE_1, STRING_BAG_ADDR);
+		BagState<String> resBag = restoredState.state(NAMESPACE_1, STRING_BAG_ADDR);
+
+		Iterator<String> orIt = orBag.get().read().iterator();
+		Iterator<String> resIt = resBag.get().read().iterator();
+
+		while (orIt.hasNext() && resIt.hasNext()) {
+			comp &= orIt.next().equals(resIt.next());
+		}
+
+		return !((orIt.hasNext() && !resIt.hasNext()) || (!orIt.hasNext() && resIt.hasNext())) && comp;
+	}
+
+	private FlinkStateInternals<String> createState(String key) throws CannotProvideCoderException {
+		return new FlinkStateInternals<>(
+				key,
+				StringUtf8Coder.of(),
+				IntervalWindow.getCoder(),
+				combiner.asKeyedFn());
+	}
+
+	@Test
+	public void test() throws Exception {
```

Comment:
Just my personal opinion. This test could have been broken into multiple tests. The init and teardown code could have been put in `@Before` and `@After` methods. 

To me, it seems much easier for debugging to do an `assertTrue` instead of ANDing all the checks.